### PR TITLE
feat(PopoverContainer): Decoupled Positioning Logic [GM-222]

### PR DESCRIPTION
**Pull request by codecaaron**
### Overview

To ensure we can scale our use of Popovers in more scenarios like `MenuButton` `Coachmark` and other components this PR aims to decouple the positioning and popover logic into a standalone component.   Further increase the amount of options for position and alignment to fit future use cases.

**Growing Pains**
1. Popovers are not content agnostic like Overlays, meaning that content cannot be completely arbitrary.
2. Fixed positioning requires constant updates to styles and is jumpy.
3. No ability to render inline means that all popover content must be on top of our header.
4. Positioning options that are relative to the target are limited leaving only a static offset for achieving certain positions.  This is prone to breakage as elements resize.

**Proposed Changes**
1. Introduce `PopoverContainer` that solely positions itself and its content.
2. Add the ability to render a Popover inline and position relative to the offsetParent rather than the viewport.
3. Use only position absolute for popover outlets to ensure that scrolling and resizing do not result in unnecessary rerenders / jumpy popovers.
4. Rethink the Popover API to allow for more direct control over how the popover is positioned.
- `inline`:  Removes the body portal and changes the calculation of where to position from to the nearest offset parent.  This allows for popovers to be inside the DOM that they target in a more accessible and predictable way. 
- `alignment`: Position the popover at the exact edge of these coordinates.
- `offset`: Addition pixel offset from center in the direction of the alignment (e.g. `top-left` and offset 20 results in 20 pixels to the left/top).
- `insideAxis`:  Flip the popover content along either the `x` or `y` axis relative to the alignment.  `top-left` and `x` would render the popover on the inside left edge rather than outside. 
- `x` | `y`:  Absolute offset values irrespective of the relative position.  (e.g. `x={-4}` will be 4 pixels to the left in every position unlike offset).

**Usage**
```tsx
const MyApp = () => {
  const target = useRef<HTMLButtonElement>(null);

  return (
    <Box>
      <Box position="relative">
        <FillButton ref={target} />
        <PopoverContainer inline targetRef={target} alignment="left">
          I render relative to my nearest parent!
        </PopoverContainer>
      </Box>
      <PopoverContainer targetRef={target} alignment="left">
        I render relative to the window!
      </PopoverContainer>
    </Box>
  );
};
```

Alignments + Transform Axis Screenshot:
Screen Shot 2021-05-10 at 9 44 08 AM

### PR Checklist

- [ ] Related to designs:
- [x] Related to JIRA ticket: GM-222
- [x] I have run this code to verify it works
- [x] This PR includes unit tests for the code change

### CODE REVIEW

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6092c2af4fadbb1b889dfe5d--gamut-preview.netlify.app>https://6092c2af4fadbb1b889dfe5d--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6092c2af4fadbb1b889dfe5d>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.3.79-alpha.c12b64.0<br/>@codecademy/gamut-labs@15.5.3-alpha.c12b64.0<br/>@codecademy/gamut@32.12.1-alpha.c12b64.0<br/>@codecademy/styleguide@36.14.1-alpha.c12b64.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6092c89f7b4eac227ae077d4--gamut-preview.netlify.app>https://6092c89f7b4eac227ae077d4--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6092c89f7b4eac227ae077d4>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.3.79-alpha.1154ad.0<br/>@codecademy/gamut-labs@15.5.3-alpha.1154ad.0<br/>@codecademy/gamut@32.12.1-alpha.1154ad.0<br/>@codecademy/styleguide@36.14.1-alpha.1154ad.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6092cd0190216d20735c68d3--gamut-preview.netlify.app>https://6092cd0190216d20735c68d3--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6092cd0190216d20735c68d3>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.3.87-alpha.e4233f.0<br/>@codecademy/gamut-labs@15.7.2-alpha.e4233f.0<br/>@codecademy/gamut@32.15.2-alpha.e4233f.0<br/>@codecademy/styleguide@37.1.3-alpha.e4233f.0<br/></pre>

**Review by sylvanasantos — COMMENTED** [review]
Jeez this is a lot of work for this popover lol, but love the improvements! And thanks for all the tests! Are you planning to make the story more interactive?

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.75-alpha.5f5843.0<br/>@codecademy/gamut-labs@19.26.4-alpha.5f5843.0<br/>@codecademy/gamut@37.3.3-alpha.5f5843.0<br/>@codecademy/monaco@0.4.3-alpha.5f5843.0<br/>@codecademy/styleguide@44.4.1-alpha.5f5843.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://614099a9e8fc4b25d83bb171--gamut-preview.netlify.app>https://614099a9e8fc4b25d83bb171--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/614099a9e8fc4b25d83bb171>Deploy Logs</a></p>

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.75-alpha.bcfb7e.0<br/>@codecademy/gamut-labs@19.26.4-alpha.bcfb7e.0<br/>@codecademy/gamut@37.3.3-alpha.bcfb7e.0<br/>@codecademy/monaco@0.4.3-alpha.bcfb7e.0<br/>@codecademy/styleguide@44.4.1-alpha.bcfb7e.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://6140a2328b4ab7447775d84c--gamut-preview.netlify.app>https://6140a2328b4ab7447775d84c--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/6140a2328b4ab7447775d84c>Deploy Logs</a></p>

File: packages/styleguide/stories/Atoms/PopoverContainer/examples.tsx [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+  'bottom-left',
+] as const;
+
+export const PopoverTemplate: React.FC<PopoverContainerProps> = (args) => {
+  const target = useRef<HTMLDivElement>(null);
+
+  return (
+    <FlexBox minHeight="480px" width={1} position="relative">
+      <FlexBox center flex={1}>
+        <PopoverContainer isOpen inline targetRef={target} {...args}>
```
Comment (dreamwasp) [inline packages/styleguide/stories/Atoms/PopoverContainer/examples.tsx:25]: *chefs kiss*

**Review by dreamwasp — APPROVED** [review]
live laugh love it

**Comment by codecademydev** [comment]
<h3>📬Published Alpha Packages:</h3><pre>@codecademy/gamut-kit@0.5.104-alpha.6b6bdc.0<br/>@codecademy/gamut-labs@24.4.2-alpha.6b6bdc.0<br/>@codecademy/gamut@37.9.1-alpha.6b6bdc.0<br/>@codecademy/monaco@0.4.16-alpha.6b6bdc.0<br/>@codecademy/styleguide@47.4.2-alpha.6b6bdc.0<br/></pre>

**Comment by codecademydev** [comment]
<p>🚀 Styleguide deploy preview ready!</p><p><a href=https://615c943ce9657d221c264dc9--gamut-preview.netlify.app>https://615c943ce9657d221c264dc9--gamut-preview.netlify.app</a></p><p><a href=https://app.netlify.com/sites/gamut-preview/deploys/615c943ce9657d221c264dc9>Deploy Logs</a></p>

**Pull request merged**